[BUG] v1.8.0 RC1 Bug - Error uploading documents (related to PaperlessTask)

### Description

When uploading multiple documents through the web ui, some documents will produce an error in the post_document API endpoint

### Steps to reproduce

1. Upload a number of documents through web ui
2. observe logs for the below

### Webserver logs

```bash
paperless-ngx-dev-webserver  | [2022-07-18 08:10:37,986] [ERROR] [django.request] Internal Server Error: /api/documents/post_document/
paperless-ngx-dev-webserver  | Traceback (most recent call last):
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 657, in get_or_create
paperless-ngx-dev-webserver  |     return self.get(**kwargs), False
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 496, in get
paperless-ngx-dev-webserver  |     raise self.model.DoesNotExist(
paperless-ngx-dev-webserver  | documents.models.PaperlessTask.DoesNotExist: PaperlessTask matching query does not exist.
paperless-ngx-dev-webserver  |
paperless-ngx-dev-webserver  | During handling of the above exception, another exception occurred:
paperless-ngx-dev-webserver  |
paperless-ngx-dev-webserver  | Traceback (most recent call last):
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
paperless-ngx-dev-webserver  |     return self.cursor.execute(sql, params)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/sqlite3/base.py", line 477, in execute
paperless-ngx-dev-webserver  |     return Database.Cursor.execute(self, query, params)
paperless-ngx-dev-webserver  | sqlite3.OperationalError: database is locked
paperless-ngx-dev-webserver  |
paperless-ngx-dev-webserver  | The above exception was the direct cause of the following exception:
paperless-ngx-dev-webserver  |
paperless-ngx-dev-webserver  | Traceback (most recent call last):
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 472, in thread_handler
paperless-ngx-dev-webserver  |     raise exc_info[1]
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 42, in inner
paperless-ngx-dev-webserver  |     response = await get_response(request)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 253, in _get_response_async
paperless-ngx-dev-webserver  |     response = await wrapped_callback(
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 435, in __call__
paperless-ngx-dev-webserver  |     ret = await asyncio.wait_for(future, timeout=None)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/asyncio/tasks.py", line 442, in wait_for
paperless-ngx-dev-webserver  |     return await fut
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/asgiref/current_thread_executor.py", line 22, in run
paperless-ngx-dev-webserver  |     result = self.fn(*self.args, **self.kwargs)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/asgiref/sync.py", line 476, in thread_handler
paperless-ngx-dev-webserver  |     return func(*args, **kwargs)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
paperless-ngx-dev-webserver  |     return view_func(*args, **kwargs)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 84, in view
paperless-ngx-dev-webserver  |     return self.dispatch(request, *args, **kwargs)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
paperless-ngx-dev-webserver  |     response = self.handle_exception(exc)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
paperless-ngx-dev-webserver  |     self.raise_uncaught_exception(exc)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
paperless-ngx-dev-webserver  |     raise exc
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
paperless-ngx-dev-webserver  |     response = handler(request, *args, **kwargs)
paperless-ngx-dev-webserver  |   File "/usr/src/paperless/src/documents/views.py", line 544, in post
paperless-ngx-dev-webserver  |     async_task(
paperless-ngx-dev-webserver  |   File "/usr/src/paperless/src/docker/src/django-q/django_q/tasks.py", line 67, in async_task
paperless-ngx-dev-webserver  |     pre_enqueue.send(sender="django_q", task=task)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 176, in send
paperless-ngx-dev-webserver  |     return [
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
paperless-ngx-dev-webserver  |     (receiver, receiver(signal=self, sender=sender, **named))
paperless-ngx-dev-webserver  |   File "/usr/src/paperless/src/documents/signals/handlers.py", line 509, in init_paperless_task
paperless-ngx-dev-webserver  |     paperless_task, created = PaperlessTask.objects.get_or_create(
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
paperless-ngx-dev-webserver  |     return getattr(self.get_queryset(), name)(*args, **kwargs)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 664, in get_or_create
paperless-ngx-dev-webserver  |     return self.create(**params), True
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 514, in create
paperless-ngx-dev-webserver  |     obj.save(force_insert=True, using=self.db)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 806, in save
paperless-ngx-dev-webserver  |     self.save_base(
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 857, in save_base
paperless-ngx-dev-webserver  |     updated = self._save_table(
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 1000, in _save_table
paperless-ngx-dev-webserver  |     results = self._do_insert(
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 1041, in _do_insert
paperless-ngx-dev-webserver  |     return manager._insert(
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
paperless-ngx-dev-webserver  |     return getattr(self.get_queryset(), name)(*args, **kwargs)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 1434, in _insert
paperless-ngx-dev-webserver  |     return query.get_compiler(using=using).execute_sql(returning_fields)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1621, in execute_sql
paperless-ngx-dev-webserver  |     cursor.execute(sql, params)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute
paperless-ngx-dev-webserver  |     return self._execute_with_wrappers(
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
paperless-ngx-dev-webserver  |     return executor(sql, params, many, context)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
paperless-ngx-dev-webserver  |     return self.cursor.execute(sql, params)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 91, in __exit__
paperless-ngx-dev-webserver  |     raise dj_exc_value.with_traceback(traceback) from exc_value
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute
paperless-ngx-dev-webserver  |     return self.cursor.execute(sql, params)
paperless-ngx-dev-webserver  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/sqlite3/base.py", line 477, in execute
paperless-ngx-dev-webserver  |     return Database.Cursor.execute(self, query, params)
paperless-ngx-dev-webserver  | django.db.utils.OperationalError: database is locked
```


### Paperless-ngx version

beta rc1

### Host OS

Ubuntu 20.04

### Installation method

Docker - official image

### Browser

_No response_

### Configuration changes

_No response_

### Other

_No response_

--- a/docs/configuration.rst
+++ b/docs/configuration.rst
@@ -31,7 +31,7 @@ PAPERLESS_REDIS=<url>
 
 PAPERLESS_DBHOST=<hostname>
     By default, sqlite is used as the database backend. This can be changed here.
-    Set PAPERLESS_DBHOST and PostgreSQL will be used instead of mysql.
+    Set PAPERLESS_DBHOST and PostgreSQL will be used instead of sqlite.
 
 PAPERLESS_DBPORT=<port>
     Adjust port if necessary.
@@ -59,6 +59,13 @@ PAPERLESS_DBSSLMODE=<mode>
     See `the official documentation about sslmode <https://www.postgresql.org/docs/current/libpq-ssl.html>`_.
 
     Default is ``prefer``.
+
+PAPERLESS_DB_TIMEOUT=<float>
+    Amount of time for a database connection to wait for the database to unlock.
+    Mostly applicable for an sqlite based installation, consider changing to postgresql
+    if you need to increase this.
+
+    Defaults to unset, keeping the Django defaults.
 
 Paths and folders
 #################

--- a/docs/troubleshooting.rst
+++ b/docs/troubleshooting.rst
@@ -309,7 +309,7 @@ You might find messages like these in your log files:
 
 .. code::
 
-    [WARNING] [paperless.management.consumer] Not consuming file /usr/src/paperless/src/../consume/SCN_0001.pdf: OS reports file as busy still
+    [ERROR] [paperless.management.consumer] Creating PaperlessTask failed: db locked
 
 You are likely using an sqlite based installation, with an increased number of workers and are running into sqlite's concurrency limitations.
 Uploading or consuming multiple files at once results in many workers attempting to access the database simultaneously.

--- a/docs/troubleshooting.rst
+++ b/docs/troubleshooting.rst
@@ -301,3 +301,19 @@ try adjusting the :ref:`polling configuration <configuration-polling>`.
 
     The user will need to manually move the file out of the consume folder and
     back in, for the initial failing file to be consumed.
+
+Log reports "Creating PaperlessTask failed".
+#########################################################
+
+You might find messages like these in your log files:
+
+.. code::
+
+    [WARNING] [paperless.management.consumer] Not consuming file /usr/src/paperless/src/../consume/SCN_0001.pdf: OS reports file as busy still
+
+You are likely using an sqlite based installation, with an increased number of workers and are running into sqlite's concurrency limitations.
+Uploading or consuming multiple files at once results in many workers attempting to access the database simultaneously.
+
+Consider changing to the PostgreSQL database if you will be processing many documents at once often.  Otherwise,
+try tweaking the ``PAPERLESS_DB_TIMEOUT`` setting to allow more time for the database to unlock.  This may have
+minor performance implications.

--- a/src/documents/signals/handlers.py
+++ b/src/documents/signals/handlers.py
@@ -11,6 +11,7 @@ from django.contrib.contenttypes.models import ContentType
 from django.db import DatabaseError
 from django.db import models
 from django.db.models import Q
+from django.db.utils import OperationalError
 from django.dispatch import receiver
 from django.utils import termcolors
 from django.utils import timezone
@@ -506,21 +507,28 @@ def add_to_index(sender, document, **kwargs):
 @receiver(django_q.signals.pre_enqueue)
 def init_paperless_task(sender, task, **kwargs):
     if task["func"] == "documents.tasks.consume_file":
-        paperless_task, created = PaperlessTask.objects.get_or_create(
-            task_id=task["id"],
-        )
-        paperless_task.name = task["name"]
-        paperless_task.created = task["started"]
-        paperless_task.save()
+        try:
+            paperless_task, created = PaperlessTask.objects.get_or_create(
+                task_id=task["id"],
+            )
+            paperless_task.name = task["name"]
+            paperless_task.created = task["started"]
+            paperless_task.save()
+        except OperationalError as e:
+            logger.error(f"Creating PaperlessTask failed: {e}")
 
 
 @receiver(django_q.signals.pre_execute)
 def paperless_task_started(sender, task, **kwargs):
     try:
         if task["func"] == "documents.tasks.consume_file":
-            paperless_task = PaperlessTask.objects.get(task_id=task["id"])
+            paperless_task, created = PaperlessTask.objects.get_or_create(
+                task_id=task["id"],
+            )
             paperless_task.started = timezone.now()
             paperless_task.save()
+    except OperationalError as e:
+        logger.error(f"Creating PaperlessTask failed: {e}")
     except PaperlessTask.DoesNotExist:
         pass
 
@@ -529,8 +537,12 @@ def paperless_task_started(sender, task, **kwargs):
 def update_paperless_task(sender, instance, **kwargs):
     try:
         if instance.func == "documents.tasks.consume_file":
-            paperless_task = PaperlessTask.objects.get(task_id=instance.id)
+            paperless_task, created = PaperlessTask.objects.get_or_create(
+                task_id=instance.id,
+            )
             paperless_task.attempted_task = instance
             paperless_task.save()
+    except OperationalError as e:
+        logger.error(f"Creating PaperlessTask failed: {e}")
     except PaperlessTask.DoesNotExist:
         pass

--- a/src/paperless/settings.py
+++ b/src/paperless/settings.py
@@ -327,6 +327,13 @@ if os.getenv("PAPERLESS_DBHOST"):
     if os.getenv("PAPERLESS_DBPORT"):
         DATABASES["default"]["PORT"] = os.getenv("PAPERLESS_DBPORT")
 
+if os.getenv("PAPERLESS_DB_TIMEOUT") is not None:
+    _new_opts = {"timeout": float(os.getenv("PAPERLESS_DB_TIMEOUT"))}
+    if "OPTIONS" in DATABASES["default"]:
+        DATABASES["default"]["OPTIONS"].update(_new_opts)
+    else:
+        DATABASES["default"]["OPTIONS"] = _new_opts
+
 DEFAULT_AUTO_FIELD = "django.db.models.AutoField"
 
 ###############################################################################

